### PR TITLE
feat(appmesh): ipv6 support for app mesh

### LINKED ISSUE #20737: aws-appmesh: Add IPv6 support
### Describe the feature

AWS App Mesh has released support for IPv6 and new fields have been exposed on the APIs as part of this release.

https://aws.amazon.com/about-aws/whats-new/2022/05/aws-app-mesh-supports-ipv6/
https://github.com/aws/aws-app-mesh-roadmap/issues/314

The proposal here is to add support for App Mesh's newly added IPv6 capabilities.

### Use Case

Without this feature, CDK users will be unable to configure IPv6 within App Mesh.

### Proposed Solution

The primary changes will be to add an IP preference that covers all of the preference types supported in App Mesh and to expose the new fields where this IP preference can be applied within the mesh and virtual node resources.

#### Mesh
```
new appmesh.Mesh(stack, 'mesh', {
  meshName: 'mesh',
  meshServiceDiscovery: {
    ipPreference: appmesh.IpPreference.IPV4_PREFERRED
  },
});
```

#### Virtual Node
```
declare const mesh: appmesh.Mesh;

const vpc = new ec2.Vpc(this, 'vpc');
const namespace = new cloudmap.PrivateDnsNamespace(this, 'test-namespace', {
    vpc,
    name: 'domain.local',
});
const service = namespace.createService('service');

/**
 * CloudMap Service Discovery
 */

const cloudNode = mesh.addVirtualNode('virtual-node', {
  serviceDiscovery: appmesh.ServiceDiscovery.cloudMap(service),
  listeners: [appmesh.VirtualNodeListener.http({
    port: 8081,
    healthCheck: appmesh.HealthCheck.http({
      healthyThreshold: 3,
      interval: cdk.Duration.seconds(5),
      path: '/health-check-path',
      timeout: cdk.Duration.seconds(2),
      unhealthyThreshold: 2,
    }),
  })],
  accessLog: appmesh.AccessLog.fromFilePath('/dev/stdout'),
},
appmesh.IpPreference.IPV4_PREFERRED);

/**
 * DNS Service Discovery
 */

const dnsNode = mesh.addVirtualNode('virtual-node', {
  serviceDiscovery: appmesh.ServiceDiscovery.dns(`service`, appmesh.DnsResponseType.ENDPOINTS, appmesh.IpPreference.IPV4_PREFERRED),
  listeners: [appmesh.VirtualNodeListener.http({
    port: 8081,
    healthCheck: appmesh.HealthCheck.http({
      healthyThreshold: 3,
      interval: cdk.Duration.seconds(5),
      path: '/health-check-path',
      timeout: cdk.Duration.seconds(2),
      unhealthyThreshold: 2,
    }),
  })],
  accessLog: appmesh.AccessLog.fromFilePath('/dev/stdout'),
});
```

### Other Information

_No response_

### Acknowledgements

- [x] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.28.0

### Environment details (OS name and version, etc.)

macOS Catalina, Version 10.15.7 (19H1824)

### PR DESCRIPTION
**Pull request by AKBarcenas**
App Mesh has released IPv6 support. This has been exposed in the form of IP preferences which have been added to the Mesh and Virtual Node resources. IP preferences are optional for both resources and there is no default IP preference that is applied by App Mesh.

The following are samples of App Mesh resources with IP preferences configured.
```
# Mesh
"spec": {
    "serviceDiscovery": {
        "ipPreference": "IPv6_PREFERRED"
    }
}
```

```
# Virtual Node
"spec": {
    "listeners": [
        {
            "healthCheck": {
                "healthyThreshold": 2,
                "intervalMillis": 5000,
                "path": "/ping",
                "protocol": "http",
                "timeoutMillis": 2000,
                "unhealthyThreshold": 2
            },
            "portMapping": {
                "port": 9080,
                "protocol": "http"
            }
        }
    ],
    "serviceDiscovery": {
        "dns": {
            "hostname": "colorteller-red.default.svc.cluster.local",
            "ipPreference": "IPv4_ONLY"
        }
    }
}
```

IP preferences on a Mesh apply the preference to all Virtual Nodes contained within that Mesh. IP preferences set on a Virtual Node will only apply to that particular Virtual Node. Additionally, Virtual Node IP preferences will override the Mesh IP preference if there is one present.

There are three areas in which the IP preference impacts how Envoy configuration generation. Firstly, setting any IP preference will change the Envoy's listeners (ingress and egress) to bind to IPv4 and IPv6 allowing the Envoy to serve all traffic from both IP versions. Secondly, the IP version specified in the name of the preference will be the IP version used for sending traffic to the local application for Envoys running as a sidecar to an application. (IPv4_ONLY/PREFERRED - IPv4, IPv6_ONLY/PREFERRED - IPv6) Lastly, it will impact how each service discovery option will be treated. For CloudMap service discovery, ONLY options will only return IPs from CloudMap for the matching version type and PREFERRED options will first used the primary IP version first and fall back to the other IP version for the IPs returned from CloudMap. For DNS service discovery, it will be similar to CloudMap service discovery in terms of only using one IP version or fall back behavior. However, this will come in the form of changing the Envoy's DNS resolver to exhibit this behavior when performing DNS resolution.

This is a summarized version of the feature. For more details, a more thorough write up can be found here: https://github.com/aws/aws-app-mesh-examples/tree/main/walkthroughs/howto-ipv6#ip-preferences-in-meshes-and-virtual-nodes

Closes #20737

### All Submissions:

* [Y] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [N] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [Y] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [Y] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by comcalvi — CHANGES_REQUESTED** [review]
Please update this PR's description to describe the feature you're adding and give a high-level overview of how it works.

File: packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts [hunk 1]
```diff
@@ -61,9 +78,13 @@ export abstract class ServiceDiscovery {
    *  filter instances by any custom attribute that you specified when you
    *  registered the instance. Only instances that match all of the specified
    *  key/value pairs will be returned.
+   * @param ipPreference Specifies the IP preference to apply to the virtual node.
+   *  The default is no IP preference being applied. No IP preference configures Envoy listeners to bind only to IPv4,
+   *  configures the Envoy to use IPv4 when sending traffic to a local application,
+   *  and configures App Mesh to prefer using IPv4 and fall back to IPv6 for IPs returned by CloudMap.
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts:84]: these lines (only 82-84) should be removed.

File: packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts [hunk 2]
```diff
@@ -48,9 +61,13 @@ export abstract class ServiceDiscovery {
    * @param hostname
    * @param responseType Specifies the DNS response type for the virtual node.
    *  The default is `DnsResponseType.LOAD_BALANCER`.
+   * @param ipPreference Specifies the IP preference to apply to the virtual node.
+   *  The default is no IP preference being applied. No IP preference configures Envoy listeners to bind only to IPv4,
+   *  configures the Envoy to use IPv4 when sending traffic to a local application,
+   *  and configures the Envoy DNS resolver to prefer using IPv6 and fall back to IPv4.
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts:67]: these 3 lines (65-67) should be moved above.

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts [hunk 3]
```diff
@@ -29,7 +34,7 @@ const virtualService = new appmesh.VirtualService(stack, 'service', {
 });
 
 const node = mesh.addVirtualNode('node', {
-  serviceDiscovery: appmesh.ServiceDiscovery.dns(`node1.${namespace.namespaceName}`),
+  serviceDiscovery: appmesh.ServiceDiscovery.dns(`node1.${namespace.namespaceName}`, undefined, appmesh.IpPreference.IPV4_ONLY),
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:37]: how does the IP preference of the `serviceDiscovery` in this node interact with the ip preference of the `serviceDiscovery` set on lines 21-23? Does one overwrite the other? Is this allowed? This behavior should be documented in the comment block of the method.

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts [hunk 4]
```diff
@@ -980,6 +982,69 @@ describe('virtual node', () => {
           },
         });
       });
+
+      test('has an IP Preference applied', () => {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:986]: we're missing a test that verifies that the ip preference is not set by default.

**Review by comcalvi — DISMISSED** [review]
This is a really strong start. Nice work!

File: packages/@aws-cdk/aws-appmesh/lib/ip-preference.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Enum of supported IP preferences.
+ * Used to dictate the IP version for mesh wide and virtual node service discovery.
+ * Also used to specify the IP version that a sidecar Envoy uses when sending traffic to a local application.
+ */
+
+export enum IpPreference {
```
[the remark this answers is not in the document]
Comment (AKBarcenas) [reply packages/@aws-cdk/aws-appmesh/lib/ip-preference.ts:7]: I think that is fair. `service-discovery.ts` is also the only place it is used.

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts (hunk 3 above)
Comment (AKBarcenas) [reply packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:37]: > Please update this PR's description to describe the feature you're adding and give a high-level overview of how it works.

I've gone ahead and added some context. If further elaboration would help, feel free to let me know.

Comment (AKBarcenas) [reply packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:37]: There are two resource levels we allowing setting this preference as you point out. A Mesh is the top level resource which users will be creating Virtual Nodes within. A Mesh preference will apply to all Virtual Nodes as a default value and each individual Virtual Node can set its own preference to override the Mesh preference. If there is no preference set anywhere then no default preference is applied.

File: packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts [hunk 6]
```diff
@@ -1,6 +1,19 @@
...
+import { IpPreference } from './ip-preference';
+
+/**
+ * Properties for Mesh Service Discovery
+ */
+export interface MeshServiceDiscovery {
+  /**
+   * IP preference applied to all Virtual Nodes in the Mesh
+   *
+   * @default No IP preference is applied to Virtual Nodes
```
[the remark this answers is not in the document]
Comment (AKBarcenas) [reply packages/@aws-cdk/aws-appmesh/lib/service-discovery.ts:13]: I've gone ahead and consolidated everything. Let me know if there is further feedback on this.

File: packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts (hunk 4 above)
Comment (AKBarcenas) [reply packages/@aws-cdk/aws-appmesh/test/virtual-node.test.ts:986]: There is another test which implicitly tests this, but I've gone ahead and added an explicit test.

**Review by comcalvi — APPROVED** [review]

**Pull request merged**